hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -756,25 +756,29 @@ private void connectToClusterInternal() {
         int attempt = 0;
         Set<InetSocketAddress> triedAddresses = new HashSet<InetSocketAddress>();
 
-        mainLoop:
         while (attempt < connectionAttemptLimit) {
             attempt++;
             long nextTry = Clock.currentTimeMillis() + connectionAttemptPeriod;
 
             Collection<InetSocketAddress> socketAddresses = getSocketAddresses();
             for (InetSocketAddress inetSocketAddress : socketAddresses) {
                 if (!client.getLifecycleService().isRunning()) {
-                    if (logger.isFinestEnabled()) {
-                        logger.finest("Giving up on retrying to connect to cluster since client is shutdown.");
-                    }
-                    break mainLoop;
+                    throw new IllegalStateException("Giving up on retrying to connect to cluster since client is shutdown.");
                 }
                 triedAddresses.add(inetSocketAddress);
                 if (connectAsOwner(inetSocketAddress) != null) {
                     return;
                 }
             }
 
+            /**
+             * If the address providers load no addresses (which seems to be possible), then the above loop is not entered
```

Comment:
Is there a way to have empty socketAddres list ?
If everything is empty we are are trying localhost right ? 
So there is always one member in the list.